### ISSUE TITLE
Have a (private?) function for each method and check the input values

### ISSUE THREAD

**Issue by BoxiLi** [flowerthrower/qutip-qoc#29]
The `optimize_pulses` method now is very general, similar to the `scipy.optimize.minimize`.  Unfortunately, different control methods have very different input formats, e.g. GRAPE takes no initial pulse guess, CRAB's pulse guess is the full pulse amplitude array, GOAT and JOPT need initial parameters for the analytical pulse function. There is no way to unify them. The input value is not sufficiently checked right now for `optimize_pulse`, input values get ignored and not passed to the algorithm. It is better to check those in each individual control function.

**Comment by rochisha0** [comment]
@BoxiLi @flowerthrower I would like to work on this issue. Can you explain what needs to be done?

**Comment by BoxiLi** [comment]
Agreed. I think the central question is, how much common APIs do those methods share? A well-designed function is like `scipy.minimize`, most methods share the same input and additional parameters. A few are different, but they need to be documented clearly. Right now, I feel like the different methods like GOAT and JOPT take a completely different set of input compared to GRAPE. Maybe we should have them each as an individual function and have them optimized first. Then, have a look if we can create an abstract layer among all methods (I am still not sure if we should do that), and group some of the input parameters as config keyword arguments.

> Moreover, the user has to consult qutip_qtrl.pulseoptim.optimize_pulse for algorithm-specific arguments

This should not happen. We have AI now. May be doc can be done much more efficiently.

> P.S. Regarding the labeling of the issue: I don't think it's a good first issue to work on since it is mostly about design choices for qutip-qoc and not a small fix / feature.

Label removed.
